Add AWS and MkDocs example; add devcontainer configuration

shepherdjerred opened the pull request:
This PR adds examples showing how to use Earthly with AWS OIDC/SSO, and MkDocs with live reloading. In addition, I added a basic devcontainer config with Docker + Earthly suitable for use with GitHub Codespaces.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/earthly/earthly?pullRequest=2859) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/earthly/earthly?pullRequest=2859) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/earthly/earthly?pullRequest=2859) it.</sub>

alexcb commented on examples/aws-sso/Earthfile:
```diff
@@ -0,0 +1,85 @@
+VERSION 0.7
+FROM amazon/aws-cli
+ARG --global EARTHLY_CI
```
should this default to `false` rather than the empty string?
```suggestion
ARG --global EARTHLY_CI=false
```

shepherdjerred replied:
I've had no problem with this being empty, but I have no problem setting it to false if you think that's better.

alexcb commented on examples/aws-sso/Earthfile:
```diff
@@ -0,0 +1,85 @@
+LOGIN:
+  COMMAND
+  ARG --required sso_region
+  BUILD +login --sso_region=$sso_region
```
is the `BUILD +login` required here? I see the `COPY` below references the same target. an explicit `BUILD` is typically only required for saving an artifact to the host (`AS LOCAL`) or pushing an image.

shepherdjerred replied:
+dev.login will save a local artifact `cache/` containing the AWS credentials, which prevents the need to log in again when the layer cache is invalidated, e.g. if you're changing the login code above you, or if one other base images get updated, you don't want to have to log in again. Because of this we need to use `BUILD` all the way down :/

alexcb replied:
ah ha! I missed that part. Thanks for the explanation.

alexcb commented on examples/aws-sso/Earthfile:
```diff
@@ -0,0 +1,85 @@
+    COPY +dev.login/aws /root/.aws
+    RUN aws-export-credentials --profile default --credentials-file-profile default
+    SAVE ARTIFACT /root/.aws/credentials /credentials
+  ELSE
```
Should this validate `EARTHLY_CI` to be either `true` or `false`
```
ELSE IF [ "$EARTHLY_CI" = "true" ]

ELSE
  RUN echo "bad value for EARTHLY_CI" && exit 1
END
```

shepherdjerred replied:
My assumption was that nobody would call this by setting `EARTHLY_CI` and would rather be using the `--ci` flag, which would mean `EARTHLY_CI` is always `true` or `false`. I can add explicit checking if you think it's useful.

alexcb reviewed (commented):
Thanks for the additions!

Can you add an entry to the `+base` target of each example under https://github.com/earthly/earthly/blob/main/Earthfile#L718

i.e.
```
examples2:
  ...
  BUILD ./examples/aws-sso+base
  BUILD ./examples/mkdocs+base
```

that way we'll at least validate the examples are syntactically correct as part of our integration tests.

shepherdjerred commented on examples/mkdocs/README.md:
```diff
@@ -0,0 +1,7 @@
+# Material for MkDocs
+
+This example shows demonstrates [Material for MkDocs]() using Earthly, and includes a live reloading development enviorment with `mkdocs serve`
```
Oops forgot to put a link here

shepherdjerred commented on examples/aws-sso/.github/workflows/aws.yaml:
```diff
@@ -0,0 +1,25 @@
+        with:
+          role-to-assume: <your OIDC role ARN>
+          aws-region: <your AWS region>
+      - name: Start Development Instance
```
Forgot to change this when copy-pasting

shepherdjerred commented:
> Can you add an entry to the +base target of each example under

Which target should I call for the AWS example? Everything meaningful requires AWS creds.

Aside from that, I've addressed the feedback you gave!

alexcb commented on examples/aws-sso/Earthfile:
```diff
@@ -0,0 +1,87 @@
+
+dev.login.open:
+  ARG --required sso_region
+  LOCALLY
```
since this is `LOCALLY`, can you include a test for deps? e.g.

```
RUN command -v python3 >/dev/null || (echo "python3 is needed locally" && exit 1)
```

just in case the user doesn't have python.

alexcb commented:
> Which target should I call for the AWS example

the `+base` target (all Earthfiles have an implicit base, which is at the top of the Earthfile).

```
  BUILD ./examples/aws-sso+base
```

alexcb commented on examples/mkdocs/Earthfile:
```diff
@@ -0,0 +1,56 @@
+  LOCALLY
+  WITH DOCKER --load=+image
+    # Open a web browser once mkdocs is ready
+    RUN (curl --output /dev/null --silent --retry 30 --retry-delay 1 --retry-all-errors \
```
might be good to include a similar check here:

```
RUN command -v python3 >/dev/null || (echo "python3 is needed locally" && exit 1)
RUN command -v curl >/dev/null || (echo "curl is needed locally" && exit 1)

```

brandonSc reviewed (approved):
Thanks for sharing this! I think AWS examples will be helpful for users who are interested in doing deployments in Earthly, especially with our new CI product.

alexcb commented:
Thanks for the PR; I rebased it, made a small Earthfile change to fix the failure, and merged via 19cc25b05a9052b868e9f23ef1710d5d24b1ad6e
